// @monsantoit/superagent-oauth/lib/addAuthorization.js
const oauth = require("@monsantoit/oauth-ping")
const azureOauth = require("@monsantoit/oauth-azure")

module.exports = {
  async pingAuthorize(params) {
    const { CLIENT_ID, CLIENT_SECRET, PING_URL } = params

    if (!PING_URL || !CLIENT_ID || !CLIENT_SECRET) {
      throw new Error("PING_URL, CLIENT_ID, and CLIENT_SECRET parameters must be set")
    }

    const token = await oauth.httpGetToken({
      clientId: CLIENT_ID,
      clientSecret: CLIENT_SECRET,
      url: PING_URL,
      autoRefresh: false,
    })()
    return `Bearer ${token}`
  },

  async azureAuthorize(params) {
    const { CLIENT_ID, CLIENT_SECRET, AZURE_URL } = params
    if (!AZURE_URL || !CLIENT_ID || !CLIENT_SECRET) {
      throw new Error("AZURE_URL, CLIENT_ID, and CLIENT_SECRET parameters must be set")
    }

    const token = await azureOauth.httpGetToken({
      clientId: CLIENT_ID,
      clientSecret: CLIENT_SECRET,
      url: AZURE_URL,
      autoRefresh: false,
    })()
    return `Bearer ${token}`
  },
}
